crates/matrix-sdk/src/oidc/cross_process.rs

```diff
@@ -0,0 +1,322 @@
+        };
+
+        trace!(
+            "Hash mismatch? {:?} (prev. known={:?}, db={:?})",
+            hash_mismatch,
+            *prev_hash,
+            db_hash
+        );
+
+        let guard = CrossProcessRefreshLockGuard {
+            hash_guard: prev_hash,
+            _store_guard: store_guard,
+            hash_mismatch,
+            db_hash,
+            store: self.store.clone(),
+        };
+
+        Ok(guard)
+    }
+
+    pub async fn restore_session(&self, tokens: &SessionTokens) {
+        let prev_tokens_hash = compute_session_hash(tokens);
+        *self.known_session_hash.lock().await = Some(prev_tokens_hash);
+    }
+
+    pub async fn on_logout(&self) -> Result<(), CrossProcessRefreshLockError> {
+        self.store
+            .remove_custom_value(OIDC_SESSION_HASH_KEY)
+            .await
+            .map_err(CrossProcessRefreshLockError::StoreError)?;
+        *self.known_session_hash.lock().await = None;
+        Ok(())
+    }
+}
+
+pub(super) struct CrossProcessRefreshLockGuard {
+    /// The hash for the latest session, either we one we knew, or the latest
+    /// one read from the database, if it was more up to date.
+    hash_guard: OwnedMutexGuard<Option<SessionHash>>,
+
+    /// Cross-process lock being hold.
+    _store_guard: CrossProcessStoreLockGuard,
+
+    /// Reference to the underlying store, for storing the hash of the latest
+    /// known session (as a custom value).
+    store: Store,
+
+    /// Do the in-memory hash and database hash mismatch?
+    ///
+    /// If so, this indicates that another process may have refreshed the token
+    /// in the background.
+    ///
+    /// We don't consider it a mismatch if there was no previous value in the
+    /// database. We do consider it a mismatch if there was no in-memory
+    /// value known, but one was known in the database.
+    pub hash_mismatch: bool,
+
+    /// Session hash previously stored in the DB. Used for debugging and testing
+    /// purposes.
+    db_hash: Option<SessionHash>,
+}
+
+impl CrossProcessRefreshLockGuard {
+    /// Updates the `SessionTokens` hash in-memory only.
+    fn save_in_memory(&mut self, hash: SessionHash) {
+        *self.hash_guard = Some(hash);
+    }
+
+    /// Updates the `SessionTokens` hash in the database only.
+    async fn save_in_database(
+        &self,
+        hash: SessionHash,
+    ) -> Result<(), CrossProcessRefreshLockError> {
+        self.store.set_custom_value(OIDC_SESSION_HASH_KEY, hash.0.to_le_bytes().to_vec()).await?;
+        Ok(())
+    }
+
+    /// Updates the `SessionTokens` hash in both memory and database.
+    ///
+    /// Must be called after a successful refresh.
+    pub async fn save_in_memory_and_db(
+        &mut self,
+        tokens: &SessionTokens,
+    ) -> Result<(), CrossProcessRefreshLockError> {
+        let hash = compute_session_hash(tokens);
+        // Note: fallible operation at the end, so the in-memory value at least is up to
+        // date.
```

Comment:
Is a partial update really what we want if a full update fails?

Reply:
I've removed that comment, as I suspect this mattered at some point during the development of the PR (i.e. just logging out an `Err` result if it failed, in a detached task), but now it's not the case anymore.